### DISCUSSION TITLE
Automatic field update

### DISCUSSION THREAD

**Discussion by thomaskamalakis** [Budibase/budibase#5767] [Help] [answered]
Hi

  I am working on a project where we need to create a table row containing the student's thesis assignment and contains the displayname and the email of faculty members supervising the thesis. These are provided through a REST query from another application.

  Is there a way to select the email of a faculty member through an option picker and then have the displayname field be automatically completed in the corresponding field form?

  What we have tried is to have a data provider inside the form returning all the faculty entries from where the email field will be selected. So its 

```
-form
   -field group
      -facultyProvider
         -emailField
            -filteredProvider
               -displayName
```

the ```facultyProvider``` returns all faculty entries and the ```filteredProvider``` has ```facultyProvider``` as data with an extra filter 

```
email Equals Binding {{ Form.Fields.emailField }}
```

The ```displayName``` has a Default Value of ```return $("filteredProvider.Rows[0]["displayName")```

This however does not seem to work dynamically, i.e. if I change the value of the ```emailField``` in the form the ```filteredProvider``` data are not updated, so ```displayName``` is incorrect. Is there a way to force the provider to update? 

On a side note, the problem seem to be similar to trying to make a simple text field ```text2Field``` have the same value as another field text value ```text1Field```. For example if the default value for ```text2Field``` is ```{{ Form.Fields.text1Field }}``` how can you ensure that it will reflect the updates of ```text1Field```? I guess we are looking something like a change listener here.

Thanks a lot

Thomas

**Comment by melohagan** [comment]
Hi Thomas, 

Is this syntax correct? `return $("filteredProvider.Rows[0]["displayName")`

You could try adding the references after; something like: `return $("Filtered Provider.Rows")[0]['displayName']`

As for forcing the data provider to update, there is a **Refresh Data Provider** action

**Comment by thomaskamalakis** [comment]
Dear @melohagan 

 Yes you are right, sorry I have mistyped the syntax when transferring it from the application. It was actually 

```
return $("filteredProvider.Rows")[0]["displayName"]
```

  to begin with. Regarding the ```Refresh Data Provider``` action I have seen this only in buttons, right? I was wondering if there is something to make the provider refresh when the ```emailField``` changes...

  In any case I did add a button ```Update``` and also a paragraph element ```testParagraph``` with its ```Text``` also set to 
```
return $("filteredProvider.Rows")[0]["displayName"]
```

inside the ```filteredProvider``` so the structure now look like this:
```
-form
   -field group
      -facultyProvider
         -emailField
            -filteredProvider
               -displayName
               -Update
               -testParagraph
```

When ever I change the ```emailField``` and click on ```Update``` the ```testParagraph``` changes to the correct displayName but the ```displayName``` field does not change value - it remains equal to the displayName of the first row returned by the outer ```facultyProvider```

Thanks!

Thomas

**Comment by melohagan** [comment]
Thanks for the update @thomaskamalakis 

The default value is not dynamic. It only applies to the initial page load.

Nonetheless, with the latest release you can now apply actions to On Change of your options picker, and use the **Update Field Value** action.

![Screenshot 2022-05-09 at 10 27 34](https://user-images.githubusercontent.com/101575380/167381579-fd5f489a-6871-4120-ad5d-8d2664ff4afd.png)

![Screenshot 2022-05-09 at 10 30 02](https://user-images.githubusercontent.com/101575380/167382027-93c2dc42-633a-4cc4-997b-4fa96f6883cb.png)

**Comment by thomaskamalakis** [comment]
Dear @melohagan 

Thank you for your suggestions. I have update the budi installation to 1.0.145 and started playing around with the ```On Change``` actions which is a very nice feature to have. My tree looks like this:
![image](https://user-images.githubusercontent.com/25413968/167422713-af69045d-8148-4043-bcd2-99ae07804f48.png)

and I have just defined the following action when the field options field ```email``` changes:

![image](https://user-images.githubusercontent.com/25413968/167423090-dc415caa-f08e-439a-92f6-51a65f9728e1.png)

```displayName``` is a simple text field. This would ordinarily simply copy the ```email``` field value to the ```displayName``` field value. However it seems to want me to select the ```email``` field twice before it updates the ```displayName```

So let's say we start from:
 
![image](https://user-images.githubusercontent.com/25413968/167423581-65cf72ed-cbca-4a36-8509-c5cf9fcfcdb6.png)

First time I select a new e-mail it goes to:

![image](https://user-images.githubusercontent.com/25413968/167423748-394d41fe-b172-4f91-a1ae-3c7e7debab56.png)

and then if I reselect the same value it then updates correctly:

![image](https://user-images.githubusercontent.com/25413968/167424076-31b5e6c7-8410-4f66-bf4b-b626e02d47b3.png)

I think the ```displayName``` field lags one step behind.

BR

Thomas

**Reply by melohagan** [reply]
Hmm that's strange, I can't reproduce that. Would you be able to share your app export if there's no confidential data?

**Comment by thomaskamalakis** [comment]
Dear @melohagan 

  I have made a simpler app exhibiting the same behavior which is OK to share (I hope) and you can find it [here](https://drive.google.com/file/d/1pgpi6_NDctUhIHy83_vnjrILKZuAUc2m/view?usp=sharing)

  Actually please ask permission to access the share.
  
BR

  Thomas

**Comment by melohagan** [comment, marked as answer]
@thomaskamalakis Thank you for sharing the app export, I see what's wrong now.

For the onChange, you had the binding `{{ Form.Fields.facultyMail }}`
Replace this with `{{ Field Value }}` - it should be at the top of the bindings list.

*Field Value* represents that value being changed to, whereas when you access the form field, it takes the value before the change occurs.

**Reply by thomaskamalakis** [reply]
Many thanks @melohagan 

Thinking of teaching budibase to our students in the conversion master programme

**Reply by melohagan** [reply]
That sounds really interesting! Which course is that? If I can be of any help don't hesitate to reach out!

**Reply by thomaskamalakis** [reply]
We should be starting a new conversion MSc in IT for people not necessarily having an IT background in 2023. We are currently thinking that budibase would be a very nice starting point for learning about databases, views, automations, APIs etc without having to hard-code everything from the get-go. Not yet sure how it will be incorporated in the syllabus but thank you very much for your kind offer.
